hack/generator/pkg/codegen/pipeline_convert_allof_and_oneof_to_objects_test.go

```diff
@@ -0,0 +1,296 @@
+
+	expected := defineEnum("b")
+
+	g.Expect(synth.intersectTypes(enum1, enum2)).To(Equal(expected))
+	g.Expect(synth.intersectTypes(enum2, enum1)).To(Equal(expected))
+}
+
+// merging objects with common properties merges the types of those properties
+func TestMergeObjectObjectCommonProperties(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	obj1 := astmodel.NewObjectType().WithProperties(
+		astmodel.NewPropertyDefinition("x", "x", defineEnum("a", "b")))
+
+	obj2 := astmodel.NewObjectType().WithProperties(
+		astmodel.NewPropertyDefinition("x", "x", defineEnum("b", "c")))
+
+	expected := astmodel.NewObjectType().WithProperties(
+		astmodel.NewPropertyDefinition("x", "x", defineEnum("b")))
+
+	g.Expect(synth.intersectTypes(obj1, obj2)).To(Equal(expected))
+	g.Expect(synth.intersectTypes(obj2, obj1)).To(Equal(expected))
+}
+
+// merging resources merges their spec & status
+func TestMergeResourceResource(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	r1 := astmodel.NewResourceType(defineEnum("a", "b"), defineEnum("x", "y"))
+	r2 := astmodel.NewResourceType(defineEnum("b", "c"), defineEnum("y", "z"))
+
+	expected := astmodel.NewResourceType(defineEnum("b"), defineEnum("y"))
+
+	g.Expect(synth.intersectTypes(r1, r2)).To(Equal(expected))
+	g.Expect(synth.intersectTypes(r2, r1)).To(Equal(expected))
+}
+
+func TestMergeResourceMissingStatus(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	r1 := astmodel.NewResourceType(defineEnum("a", "b"), defineEnum("x", "y"))
+	r2 := astmodel.NewResourceType(defineEnum("b", "c"), nil)
+
+	expected := astmodel.NewResourceType(defineEnum("b"), defineEnum("x", "y"))
+
+	g.Expect(synth.intersectTypes(r1, r2)).To(Equal(expected))
+	g.Expect(synth.intersectTypes(r2, r1)).To(Equal(expected))
+}
+
+func TestMergeResourceWithOtherDependsOnSpecVsStatus(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	r := astmodel.NewResourceType(defineEnum("a", "b"), defineEnum("c", "d"))
+	e := defineEnum("b", "c")
+
+	synth.specOrStatus = chooseSpec
+	{
+		expected := astmodel.NewResourceType(defineEnum("b"), defineEnum("c", "d"))
+		g.Expect(synth.intersectTypes(r, e)).To(Equal(expected))
+		g.Expect(synth.intersectTypes(e, r)).To(Equal(expected))
+	}
+
+	synth.specOrStatus = chooseStatus
+	{
+		expected := astmodel.NewResourceType(defineEnum("a", "b"), defineEnum("c"))
+		g.Expect(synth.intersectTypes(r, e)).To(Equal(expected))
+		g.Expect(synth.intersectTypes(e, r)).To(Equal(expected))
+	}
+
+	synth.specOrStatus = ""
+}
+
+// merging a oneOf with a type that is in the oneOf results in that type
+func TestMergeOneOf(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	oneOf := astmodel.MakeOneOfType([]astmodel.Type{astmodel.IntType, astmodel.StringType, astmodel.BoolType})
+
+	g.Expect(synth.intersectTypes(oneOf, astmodel.BoolType)).To(Equal(astmodel.BoolType))
+	g.Expect(synth.intersectTypes(astmodel.BoolType, oneOf)).To(Equal(astmodel.BoolType))
+}
+
+// wooh!
+func TestMergeOneOfEnum(t *testing.T) {
+	// this is based on a real-world example
```

Comment:
Do you remember which one? Might be worth a reference

Reply:
will add